### DISCUSSION TITLE
How convert the correct type between ModelRc and array(vec)

### DISCUSSION THREAD

**Discussion by keith331** [slint-ui/slint#5244] [General] [answered]
I'm still learning Rust and Slint, I can set a array to slint, but get array from slint. how to covnert ModelRc(dyn Model) to array(vec) effectively. appreciate it!!
Here is my code

`export component AppWindow inherits Window {
    in-out property<int> counter: 42;
    in-out property<[bool]> checked_array: [true, true, true, false];
    callback sig-checked([bool]);
    
    VerticalBox {
            for item[idx] in checked_array: CheckBox {
            text: @tr("Item {}",idx);
            checked: root.checked_array[idx];
            toggled => {
                root.checked_array[idx] = !root.checked_array[idx];
                root.sig-checked(root.checked_array);
            }
        }
    }`

`fn main() -> Result<(), slint::PlatformError> {
    let ui = AppWindow::new()?;

    //  set property
    let initial_array = [false, false, true, false];
    ui.set_checked_array(initial_array.into());

    // get status
    ui.on_sig_checked( {
        move |output_states| {
            println!("{:#?}", output_states);
        }
    });
    ui.run()
}
`

**Comment by tronical** [comment]
I think that might work is this:

```rust
output_states.as_any().downcast_ref::<slint::VecModel<bool>>().unwrap()
```

**Reply by keith331** [reply]
Well, it is still ModelRc(dyn Model)

**Reply by tronical** [reply, marked as answer]
After the `downcast_ref()` call you have a `VecModel` reference. Here's a self-contained example:

```rust
use slint::Model;

slint::slint! {
    import { CheckBox, VerticalBox } from "std-widgets.slint";
    export component App {
        in-out property<[bool]> checked_array: [true, true, true, false];
        callback sig-checked([bool]);
        VerticalBox {
            for item[idx] in checked_array: CheckBox {
                text: @tr("Item {}",idx);
                checked: root.checked_array[idx];
                toggled => {
                    root.checked_array[idx] = !root.checked_array[idx];
                    root.sig-checked(root.checked_array);
                }
            }
        }
    }
}

fn main() -> Result<(), slint::PlatformError> {
    let app = App::new()?;

    app.on_sig_checked(|model| {
        let vecmodel: &slint::VecModel<bool> = model
            .as_any()
            .downcast_ref()
            .unwrap();
        let states = vecmodel.iter().collect::<Vec<_>>();
        eprintln!("states: {:#?}", states);
    });

    app.run()
}
```

**Reply by keith331** [reply]
It works. It's quite complicated for me. but slint is really cool. thank you so much @tronical
